src/thread_event.c

```diff
@@ -0,0 +1,277 @@
+			}						\
+		}							\
+		thread_event_wait_assert(event_wait);			\
+		event##_event_wait_set(tsd, event_wait);		\
+		/*							\
+		 * If there is a single event, then the remaining wait	\
+		 * time may become zero, and we rely on either the	\
+		 * event handle or a thread_event_update() call later	\
+		 * to properly set next_event; if there are multiple	\
+		 * events, then	here we can get the minimum remaining	\
+		 * wait time to	the next already set event.		\
+		 */							\
+		if (event_wait > 0 && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	/*
+	 * In case none of the events has its condition on, set next_event to
+	 * be THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX to minimize malloc() slow
+	 * path triggering.
+	 */
+	if (no_event_on) {
+		/* The subtraction is intentionally susceptible to underflow. */
+		wait = THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX -
+		    thread_allocated_after;
+		if (wait < THREAD_EVENT_MIN_WAIT) {
+			wait = THREAD_EVENT_MIN_WAIT;
+		}
+	}
+
+	thread_event_wait_assert(wait);
+	thread_allocated_last_event_set(tsd, thread_allocated_after);
+	thread_event_adjust_thresholds_helper(tsd, thread_allocated_after,
+	    wait, no_event_on);
+
+	thread_event_assert_invariants(tsd);
+
+#define W(event, condition)						\
+	if (condition) {						\
+		thread_##event##_event_handle(tsd);			\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	thread_event_assert_invariants(tsd);
+}
+
+void
+thread_event_rollback(tsd_t *tsd, size_t diff) {
+	thread_event_assert_invariants(tsd);
+
+	if (diff == 0) {
+		return;
+	}
+
+	uint64_t thread_allocated = thread_allocated_get(tsd);
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t thread_allocated_rollback = thread_allocated - diff;
+	thread_allocated_set(tsd, thread_allocated_rollback);
+
+	uint64_t last_event = thread_allocated_last_event_get(tsd);
+	/* Both subtractions are intentionally susceptible to underflow. */
+	if (thread_allocated_rollback - last_event <=
+	    thread_allocated - last_event) {
+		thread_event_assert_invariants(tsd);
+		return;
+	}
+
+	thread_allocated_last_event_set(tsd, thread_allocated_rollback);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t wait_diff = last_event - thread_allocated_rollback;
+	assert(wait_diff <= diff);
+	bool no_event_on = true;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		no_event_on = false;					\
+		uint64_t event_wait = event##_event_wait_get(tsd);	\
+		thread_event_wait_assert(event_wait);			\
+		if (event_wait > 0) {					\
+			if (wait_diff >					\
+			    THREAD_EVENT_MAX_WAIT - event_wait) {	\
+				event_wait = THREAD_EVENT_MAX_WAIT;	\
+			} else {					\
+				event_wait += wait_diff;		\
+			}						\
+			thread_event_wait_assert(event_wait);		\
+			event##_event_wait_set(tsd, event_wait);	\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	if (!no_event_on) { /* Double negation! */
```

Comment:
let’s compute the value of no_event_on at the end of malloc_init, and store it in a static bool. Maybe change it to thread_event_active?

Reply:
Good point. Removed most of the `no_event_on` logic, while only keeping one for periodically asserting that `thread_event_active` is still correct (in case in the future some conditions might be technically adjustable so we need to make sure that the program does not adjust them by mistake).